s390x: add support for s390x

alicefr opened the pull request:
The PR adds the support for s390x.

In the case of CCW devices, the vhost-user devices are not supported.
See #659. An error message is thrown if they tried to be used.

Memory hotplug is not supported on s390 yet and an error message is thrown.

The VirtioNetPCI has been changed to VirtioNet. The generalization
allows to set the VirtioNet to the correct CCW device for s390x.

A lot of the elements in this PR have been taken from https://github.com/kata-containers/runtime/pull/667

Fixes: #666

Co-authored-by: Yash D Jain ydjainopensource@gmail.com
Signed-off-by: Alice Frosi <afrosi@de.ibm.>

jodh-intel commented on cli/kata-check_data_s390x_test.go:
```diff
@@ -0,0 +1,21 @@
+const testCPUInfoTemplate = `
+vendor_id       : IBM/S390
+# processors    : 4
+bogomips per cpu: 20325.00
```
Wow :)

jodh-intel commented on cli/kata-check_s390x_test.go:
```diff
@@ -0,0 +1,285 @@
+	}
+}
+
+type TestDataa struct {
```
This seems to have an extra "a" at the end. But also appears to be unused?

alicefr replied:
I did the same as [`amd64`](https://github.com/kata-containers/runtime/blob/a323a87b591f2e38efb74b5b371e66f230fd188e/cli/kata-check_amd64_test.go#L487) and [`ppc64le`](https://github.com/kata-containers/runtime/blob/a323a87b591f2e38efb74b5b371e66f230fd188e/cli/kata-check_ppc64le_test.go#L211). I forgot to changed to the function `genericTestGetCPUDetails`. Now I fixed it

jodh-intel replied:
Thanks!

jodh-intel commented on cli/utils_s390x.go:
```diff
@@ -0,0 +1,12 @@
+// +build s390x
```
This is technically redundant as the filename already ends in the build architecture.

alicefr replied:
you're right. I remove it

jodh-intel commented on virtcontainers/qemu_s390x.go:
```diff
@@ -0,0 +1,123 @@
+func MaxQemuVCPUs() uint32 {
+	// Max number of virtual Cpu defined in qemu. See
+	// https://github.com/qemu/qemu/blob/80422b00196a7af4c6efb628fae0ad8b644e98af/target/s390x/cpu.h#L55
+	// #define S390_MAX_CPUS 248
```
It's a shame this can't be queried by calling qemu (via govmm) directly. But clearly not a blocker as other architectures also hard-code a value.

Aside: For the future, I do wonder if we could push the hard-coded arch-specific values down into govmm rather than having them in the runtime though - wdyt @markdryan?

jodh-intel replied:
Sure - this was more a question for @markdryan ;)

jodh-intel commented:
Ping @kata-containers/runtime! ;)

grahamwhaley commented:
/test

jodh-intel commented:
Travis is complaining due to a static analysis issue:

```
virtcontainers/qemu.go:907:9:warning: if block ends with a return statement, so drop this else and outdent its block (golint)
```

alicefr commented on virtcontainers/qemu.go:
```diff
@@ -892,9 +892,13 @@ func (q *qemu) hotplugNetDevice(endpoint Endpoint, op operation) error {
+			err = q.qmpMonitorCh.qmp.ExecuteNetPCIDeviceAdd(q.qmpMonitorCh.ctx, tap.Name, devID, endpoint.HardwareAddr(), addr, bridge.ID, romFile, int(q.config.NumVCPUs), q.arch.runNested())
 		}
+		return err
 	} else {
```
@jodh-intel I can remove this else to simplify the code structure. It's actually redundant

jodh-intel replied:
Thanks!

alicefr commented:
@jodh-intel now the changes in `virtcontainers/qemu.go` are a little bit more, but it simplify the code structure. I moved the switch out of the if statement. It was the same for both if and else statement and removed the else because the if statement has a `return err`

jodh-intel commented:
Thanks @alicefr.

jodh-intel commented:
/retest

devimc commented on virtcontainers/qemu.go:
```diff
@@ -892,33 +893,27 @@ func (q *qemu) hotplugNetDevice(endpoint Endpoint, op operation) error {
 
-		if err := q.removeDeviceFromBridge(tap.ID); err != nil {
-			return err
+		if machine, _ := q.getQemuMachine(); machine.Type == QemuCCWVirtio {
```
```go
	if machine, _ := q.getQemuMachine(); machine.Type == QemuCCWVirtio {
		return q.qmpMonitorCh.qmp.ExecuteNetCCWDeviceAdd(q.qmpMonitorCh.ctx, tap.Name, devID, endpoint.HardwareAddr(), addr, bridge.ID, int(q.config.NumVCPUs))
	}
	return q.qmpMonitorCh.qmp.ExecuteNetPCIDeviceAdd(q.qmpMonitorCh.ctx, tap.Name, devID, endpoint.HardwareAddr(), addr, bridge.ID, romFile, int(q.config.NumVCPUs), q.arch.runNested())
```

I think this looks better, @alicefr what do you think?

alicefr replied:
I changed the code as you suggested

alicefr replied:
@sboeuf you're right, I can add a check

devimc commented on virtcontainers/qemu_s390x.go:
```diff
@@ -0,0 +1,123 @@
+// appendVhostUserDevice throws an error if vhost devices are tried to be used.
+// See issue https://github.com/kata-containers/runtime/issues/659
+func (q *qemuS390x) appendVhostUserDevice(devices []govmmQemu.Device, attr config.VhostUserDeviceAttrs) []govmmQemu.Device {
+	logrus.Fatalln("No vhost-user devices supported on s390x")
```
please change `appendVhostUserDevice` in `qemuArchBase` to return `([]govmmQemu.Device, error)`

alicefr replied:
I introduced the `error` and changed in the palce where the function is called

jcvenegas commented on virtcontainers/qemu.go:
```diff
@@ -892,33 +893,25 @@ func (q *qemu) hotplugNetDevice(endpoint Endpoint, op operation) error {
+		return err
+	}
+
+	devID := "virtio-" + tap.ID
```
Nit: Now this string is repeated wonder the value make it a constant.

jodh-intel commented:
@bergwolf, @caoruidong, @grahamwhaley, @devimc - ptal

caoruidong commented:
/test

caoruidong commented on cli/kata-check_s390x.go:
```diff
@@ -0,0 +1,125 @@
+		return "", "", fmt.Errorf("cannot find vendor field in file %v", procCPUInfo)
+	}
+
+	// model name is optional
```
Is it optional?

alicefr replied:
you're right, it's not optional. I forgot to remove the comment. Before the model was not correctly parsed

caoruidong commented on virtcontainers/qemu.go:
```diff
@@ -891,34 +893,28 @@ func (q *qemu) hotplugNetDevice(endpoint Endpoint, op operation) error {
+		return err
+	}
+	if err := q.qmpMonitorCh.qmp.ExecuteNetdevDel(q.qmpMonitorCh.ctx, tap.Name); err != nil {
+		return err
```
Thanks for refactoring!

alicefr commented:
I introduced the error check for the `getQemuMachine()` and moved the `devID := "virtio-" + tap.ID
` with the switch outside the if statement. I removed the model optional comment

devimc commented:
/test

jodh-intel commented:
\o/ Thanks again @alicefr! :smile:
